---
title: 'Introduction to Bioconductor'
author: "Data Wrangling in R"
output:
  ioslides_presentation:
    widescreen: yes
---

```{r,include=FALSE}
library(stringr)
library(tibble)
library(SummarizedExperiment)
```


## The Bioconductor project

- [Bioconductor](http://www.bioconductor.org) is an open source, open development software project to provide tools for the analysis and comprehension of high-throughput genomic data. It is based primarily on the R programming language.

- Most Bioconductor components are distributed as R packages. The functional scope of Bioconductor packages includes the analysis of microarray, sequencing, flow sorting, genotype/SNP, and other data.

## Project Goals

The broad goals of the Bioconductor project are:

- To provide widespread access to a broad range of powerful statistical and graphical methods for the analysis of genomic data.
- To facilitate the inclusion of biological metadata in the analysis of genomic data, e.g. literature data from PubMed, annotation data from Entrez genes.
- To provide a common software platform that enables the rapid development and deployment of extensible, scalable, and interoperable software.
- To further scientific understanding by producing high-quality documentation and reproducible research.
- To train researchers on computational and statistical methods for the analysis of genomic data.

## Quick overview of the website

- biocViews
- Support site
- Teaching material
- Installation

## Getting started

```{r, eval=FALSE}
# Note that this is not evaluated here, so you will have to do it before using this knitr doc
install.packages("BiocManager")
# Install all core packages and update all installed packages
BiocManager::install()
```

## Getting started

You can also install specific packages

```{r, eval=FALSE}
# Note that this is not evaluated here, so you will have to do it before using this knitr doc
BiocManager::install(c("GEOquery", "limma", "biomaRt", "SummarizedExperiment"))
```

## Bioconductor Workflows

https://bioconductor.org/packages/release/workflows/vignettes/sequencing/inst/doc/sequencing.html

## The Gene Expression Omnibus (GEO)

The [Gene Expression Omnibus](http://www.ncbi.nlm.nih.gov/geo/) is an international public repository that archives and freely distributes microarray, next-generation sequencing, and other forms of high-throughput functional genomics data submitted by the research community.

The three main goals of GEO are to:

- Provide a robust, versatile database in which to efficiently store high-throughput functional genomic data
- Offer simple submission procedures and formats that support complete and well-annotated data deposits from the research community
- Provide user-friendly mechanisms that allow users to query, locate, review and download studies and gene expression profiles of interest

## Getting data from GEO

For individual studies/datasets, the easiest way to find publicly-available data is the GEO accession number found at the end of publications.

## Getting data from GEO

The `GEOquery` package can access GEO directly.

https://www.bioconductor.org/packages/release/bioc/html/GEOquery.html

```{r}
library(GEOquery)
# https://pubmed.ncbi.nlm.nih.gov/32619517/
geo_data = getGEO("GSE146760")[[1]] # find accession in paper
```

## Getting data from GEO

```{r}
tibble(pData(geo_data))
```

## Getting data from GEO

Actual gene expression data, ie RNA-seq read counts, is less commonly stored in GEO.

```{r}
exprs(geo_data) # gene expression
fData(geo_data) # gene/feature/row annotation
```

## Getting data from GEO

Sometimes the gene expression matrices are stored as supplementary data.
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE146760

```{r}
getGEOSuppFiles("GSE146760")
counts  = read.delim("GSE146760/GSE146760_RNASeq_analysis_CountsTable.txt.gz")
pheno = read.delim("GSE146760/GSE146760_RNASample_Info_v2.txt.gz")
```

## Getting data from GEO

```{r}
colnames(counts) = sapply(str_split(colnames(counts), "Aligned"), "[[", 1)
identical(colnames(counts), pheno$Prefix)
rownames(pheno) = pheno$Status
colnames(counts) = pheno$Status
```

## Getting data from GEO

SummarizedExperiment objects are probably the standard data structure for gene expression data.

https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html

```{r}
rse = SummarizedExperiment(assays = list(counts = counts),
                            colData = DataFrame(pheno))
```

## Getting data from GEO

We can also add gene annotation information with the `biomaRt` package

```{r, cache=TRUE}
library(biomaRt)
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
geneMap = getBM(attributes = c("ensembl_gene_id",
                "chromosome_name","start_position",
          "end_position", "strand", "external_gene_name"),
			values=rownames(counts), mart=ensembl)
```

## Genomic Ranges

Convert the data frame to a G[enomic]Ranges object:

```{r}
geneMap$chromosome_name = paste0("chr", geneMap$chromosome_name)
geneMap$strand = ifelse(geneMap$strand == 1, "+", "-")
geneMap_gr = makeGRangesFromDataFrame(geneMap,
            seqnames.field = "chromosome_name",
            start.field = "start_position",
            end.field = "end_position")
names(geneMap_gr) = geneMap$ensembl_gene_id
geneMap_gr
```

## Genomic Ranges

```{r}
identical(rownames(counts), names(geneMap_gr))
table(rownames(counts) %in% names(geneMap_gr))
mm = match(rownames(counts), names(geneMap_gr))
geneMap_gr = geneMap_gr[mm[!is.na(mm)]]
counts = counts[!is.na(mm),]
```

## Summarized Experiments

```{r}
rse = SummarizedExperiment(assays = list(counts = counts),
                            colData = DataFrame(pheno),
                           rowRanges = geneMap_gr)
rse
```


## Getting data from the Sequence Read Archive (SRA)

GEO originated for microarray data, which has largely become replaced by data produced using next-generation sequencing technologies. Depositing raw sequencing reads into the Sequence Read Archive (SRA) is often a condition of publication in many journals. 

https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP044749	

## Raw data is annoying to process into gene counts

So we created the `recount` project
https://jhubiostatistics.shinyapps.io/recount/

```{r}
source("scale_counts.R") # or install recount package
load(file.path('SRP044749', 'rse_gene.Rdata'))
rse_gene = scale_counts(rse_gene)
rse_gene
```

